File: flutter_application_1/lib/models/comment.dart
class Comment {
  final String id;
  final String content;
  final String status;
  final String type;
  final int score;
  final bool isEdited;
  final List<String> mentions;
  final dynamic author;
  final String? post;
  final String? answer;
  final DateTime createdAt;
  final DateTime updatedAt;

  Comment({
    required this.id,
    required this.content,
    required this.status,
    required this.type,
    required this.score,
    required this.isEdited,
    required this.mentions,
    required this.author,
    this.post,
    this.answer,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'] ?? '',
      content: json['content'] ?? '',
      status: json['status'] ?? 'approved',
      type: json['type'] ?? 'general',
      score: json['score'] ?? 0,
      isEdited: json['isEdited'] ?? false,
      mentions: List<String>.from(json['mentions'] ?? []),
      author: json['author'],
      post: json['post'],
      answer: json['answer'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'content': content,
      'type': type,
      'post': post,
      'answer': answer,
      'mentions': mentions,
    };
  }

  String get authorName {
    if (author is Map) {
      return author['username'] ?? 'Unknown';
    }
    return 'Unknown';
  }

  int get authorReputation {
    if (author is Map) {
      return author['reputation'] ?? 0;
    }
    return 0;
  }
  String? get authorId {
    if (author is String) {
      return author;
    } else if (author is Map && author['_id'] != null) {
      return author['_id'];
    }
    return null;
  }}

File: flutter_application_1_backend/src/models/Comment.js
const mongoose = require("mongoose");

const CommentSchema = new mongoose.Schema(
  {
    // String field with validation rule (minlength)
    content: {
      type: String,
      required: true,
      minlength: [3, "Comment must be at least 3 characters long"],
      maxlength: 500,
      trim: true
    },
    
    // String field with lowercase constraint
    status: {
      type: String,
      lowercase: true,
      default: "approved"
    },
    
    // String field with enum constraint
    type: {
      type: String,
      enum: {
        values: ["question", "answer", "general"],
        message: "{VALUE} is not a valid comment type"
      },
      required: true,
      default: "general"
    },
    
    // Votes array (like Posts/Answers)
    votes: [
      {
        user: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
        vote: { type: Number, enum: [1] } // Comments only support upvotes
      }
    ],
    
    // Number field with maximum value (kept for backward compatibility)
    score: {
      type: Number,
      default: 0,
      max: [1000, "Score cannot exceed 1000"],
      min: 0
    },
    
    // Boolean field
    isEdited: {
      type: Boolean,
      default: false
    },
    
    // Array field
    mentions: {
      type: [String],
      default: []
    },
    
    // JSON-like field (Object/Mixed type)
    metadata: {
      type: {
        editHistory: [{
          editedAt: Date,
          previousContent: String
        }],
        ipAddress: String,
        userAgent: String,
        flags: Number
      },
      default: {}
    },
    
    // Foreign Key to User collection
    author: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true
    },
    
    // Foreign Key to Post collection (optional - either post or answer)
    post: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Post"
    },
    
    // Foreign Key to Answer collection (optional - either post or answer)
    answer: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Answer"
    }
  },
  { 
    timestamps: true // Automatically adds createdAt and updatedAt Date fields
  }
);

// Virtual field for vote count
CommentSchema.virtual('voteCount').get(function() {
  return this.votes.length;
});

CommentSchema.set('toJSON', { virtuals: true });
CommentSchema.set('toObject', { virtuals: true });

// Index for faster queries
CommentSchema.index({ post: 1, createdAt: -1 });
CommentSchema.index({ answer: 1, createdAt: -1 });
CommentSchema.index({ author: 1 });

// Custom validation: A comment must belong to either a post or an answer
CommentSchema.path('post').validate(function(value) {
  return this.post || this.answer;
}, 'Comment must be associated with either a post or an answer');

CommentSchema.path('post').validate(function(value) {
  return !(this.post && this.answer);
}, 'Comment cannot be associated with both post and answer');

module.exports = mongoose.model("Comment", CommentSchema);
